onnxscript/rewriter/pattern.py

```diff
@@ -604,27 +570,30 @@ def __init__(
         self._rel_tol = rel_tol
         self._abs_tol = abs_tol
 
-    def match_scalar(self, scalar_value):
-        status = math.isclose(
-            scalar_value, self._value, rel_tol=self._rel_tol, abs_tol=self._abs_tol
-        )
-        # Note: If the value is produced by a Constant node, we could include
-        # the Constant node in the return_value list. However, we don't do that.
-        # Instead, we will rely on DCE to remove the constant node if it is not
-        # used elsewhere.
-        return MatchResult(success=status)
+    @property
+    def value(self) -> int | float:
+        return self._value
 
-    def matches(self, value: ir.Value):
+    def matches(self, value: ir.Value, match: MatchResult) -> MatchResult:
         value = _ir_utils.propagate_const_value(value)
         constant_value = _ir_utils.get_numpy_from_ir_value(value)
         if constant_value is None:
-            return MatchResult.FAIL()
+            return match.fail(f"Value is not a constant, expecting {self.value}.")
 
         # TODO (rama): allow users to specify shape requirement, if desired.
         if constant_value.size != 1:
-            return MatchResult.FAIL()
+            return match.fail(f"Value is not a scalar, expecting {self.value}.")
 
-        return self.match_scalar(constant_value.item())
+        if not math.isclose(
+            constant_value.item(), self._value, rel_tol=self._rel_tol, abs_tol=self._abs_tol
+        ):
+            match.fail(f"Value mismatch: expected {self._value}, got {constant_value.item()}.")
```

Comment:
Is it every possible for fail to be called multiple times on the same match result? That way the result will be overwritten.

Reply:
Not by design. The first failure terminates that particular search.